### PR TITLE
torch ddp多卡的情况下保证 Batch Norm 使用所有卡上的数据共同计算均值和方差

### PR DESCRIPTION
**Pull request by 00INDEX**
Description：torch ddp多卡的情况下保证 Batch Norm 使用所有卡上的数据共同计算均值和方差。

Main reason: 在使用 torch dpp 多卡的时候，BatchNorm 层默认值使用自己卡上的数据计算均值和方差然后进行归一化，导致不同卡上计算得到的均值和方差不同，可能会使训练变得不稳定。

Checklist  检查下面各项是否完成

Please feel free to remove inapplicable items for your PR.

-	[x] The PR title starts with [$CATEGORY] (例如[bugfix]修复bug，[new]添加新功能，[test]修改测试，[rm]删除旧代码)
-	[x] Changes are complete (i.e. I finished coding on this PR)  修改完成才提PR
-	[x] All changes have test coverage  修改的部分顺利通过测试。对于fastnlp/fastnlp/*的修改，测试代码**必须**提供在fastnlp/test/*。
-	[x] Code is well-documented  注释写好，API文档会从注释中抽取
-	[x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change  修改导致例子或tutorial有变化，请找核心开发人员

Changes: 逐项描述修改的内容
- fastNLP/core/drivers/torch_driver/dist_utils 中增加了 convert_sync_batchnorm 方法，当 ddp 得到参数 sync_bn=True (例如在 Trainer 的参数中指定 sync_bn=True) 时，可以把模型中所有的 BatchNorm 层替换成 torch 实现好的 SyncBatchNorm 层。

### CODE REVIEW

File: tests/core/drivers/torch_driver/test_ddp.py [hunk 1]
```diff
@@ -1053,4 +1053,52 @@ def __len__(self):
...
         if dist.is_initialized():
             dist.barrier()
-            dist.destroy_process_group()
\ No newline at end of file
+            dist.destroy_process_group()
+
+
+@pytest.mark.torch
+@magic_argv_env_context
+@pytest.mark.parametrize("device", ([[0, 1, 2]]))
```
Comment (yhcc) [inline tests/core/drivers/torch_driver/test_ddp.py:1061]: 这个地方修改为两张卡吧，一般多卡就测两张就可以了，不然测试环境压力太大了。同时也测试一下单卡能否正确运行不报错。

File: fastNLP/core/drivers/torch_driver/ddp.py [hunk 2]
```diff
@@ -330,6 +332,8 @@ def __init__(
 
         self._has_setup = False  # 设置这一参数是因为 evaluator 中也会进行 setup 操作，但是显然是不需要的也不应该的；
         self._has_ddpwrapped = False  # 判断传入的模型是否经过 _has_ddpwrapped 包裹；
+        # sync_bn 表示是否要将模型中可能存在的 BatchNorm 层转换为可同步所有卡数据计算均值和方差的 SyncBatchNorm 层；
+        self.sync_bn = kwargs.get("sync_bn", False)
```
Comment (yhcc) [inline fastNLP/core/drivers/torch_driver/ddp.py:336]: 我感觉应该默认为True吧？

File: fastNLP/core/drivers/torch_driver/ddp.py [hunk 3]
```diff
@@ -243,6 +244,7 @@ class TorchDDPDriver(TorchDriver):
         * *gradscaler_kwargs* -- 用于 ``fp16=True`` 时，提供给 :class:`torch.amp.cuda.GradScaler` 的参数
     :kwargs:
         * *wo_auto_param_call* (``bool``) -- 是否关闭在训练时调用我们的 ``auto_param_call`` 函数来自动匹配 batch 和前向函数的参数的行为
+        * *sync_bn* (``bool``) -- 是否要将模型中可能存在的 BatchNorm 层转换为可同步所有卡数据计算均值和方差的 SyncBatchNorm 层
```
Comment (yhcc) [inline fastNLP/core/drivers/torch_driver/ddp.py:247]: 加一个默认为True吧

**Review by yhcc — CHANGES_REQUESTED** [review]
